tsdb/engine/tsm1/compact.go

```diff
@@ -580,11 +596,26 @@ func (c *DefaultPlanner) Plan(lastWrite time.Time) ([]CompactionGroup, int64) {
 	// each generation in descending break once we see a file less than 4.
 	end := 0
 	start := 0
+	lastHighLevelIndex := -1
+
 	for i, g := range generations {
-		if g.level() <= 3 {
-			break
+		if c.enableNestedCompactor {
+			// Track the last high-level generation
+			if g.level() > 3 {
+				lastHighLevelIndex = i
+			}
+		} else {
+			if g.level() <= 3 {
+				break
+			}
+
+			end = i + 1
 		}
-		end = i + 1
+	}
+
+	// If we have high-level files, only include generations up to the last high-level file
+	if c.enableNestedCompactor && lastHighLevelIndex >= 0 {
```

Comment:
This may not be needed since we set lastHighLevelIndex to -1 so if the feature flag is disabled it will just always be -1.. I want to remain consistent and indicate that this code is only supposed to be enabled when the feature flag is on so I decided to keep it.